Move mode up to the notebook level

blink1073 opened the pull request:
Supersedes #103.

Commit:
Move mode up to the notebook level

Commit:
Remove errant variables

Commit:
Focus the editor explicitly when whe change mode

blink1073 commented:
I'm going to clean up this logic, but I've got a elementary school concert to attend first :smile:

jasongrout commented:
Have fun!

ellisonbg commented:
We shouldn't focus the editor on cell selection. The reason is that there
are many other areas of the cell a user could click on (cell toolbar,
widgets, output) that need focus. We don't want our manual focus of the
editor to have to fight with those other focus events. We have run into a
_ton_ of weird bugs anytime we manual focus an element in the notebook. In
general, there are very, very few times where it is actually the right
thing to do. Historically, command/edit mode were created, in part, to
allow us to do manual focus less often.

On Thu, Mar 17, 2016 at 10:03 AM, Jason Grout notifications@github.com
wrote:

> Have fun!
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> https://github.com/jupyter/jupyter-js-notebook/pull/105#issuecomment-197975367

## 

Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

blink1073 commented:
The manual focus is there to accomodate the "switch to edit mode" keyboard shortcut.

ellisonbg commented:
Ahh yes, as long as the manual focus is only applied in response to that
keyboard shortcut it is fine

On Thu, Mar 17, 2016 at 10:32 AM, Steven Silvester <notifications@github.com

> wrote:
> 
> The manual focus is there to accomodate the "switch to edit mode" keyboard
> shortcut.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/jupyter/jupyter-js-notebook/pull/105#issuecomment-197990090

## 

Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

Commit:
More cleanup of mode handling

blink1073 commented:
This is ready for review.  There is always an explicit focus event when a cell or the mode changes: we focus the cell node itself or the editor.  This allows us to have explicit control over the mode of a cell while having keyboard shortcuts bound to the cell CSS classes.

jasongrout commented:
Thanks.  I'm looking at it right now.

jasongrout commented on src/notebook/widget.ts:
```diff
@@ -154,6 +154,16 @@ const TOOLBAR_PRESSED = 'jp-mod-pressed';
 const TOOLBAR_BUSY = 'jp-mod-busy';
 
 /**
+ * The class name added to a cell in edit mode.
```
Now it seems clearer to me that the modes are *notebook* modes (like you call it in the model) and the class should be on the *notebook* widget. For example, we can't have some cells in edit mode and others in command mode. Instead, either the notebook as a whole is in edit mode (editing the current cell) or is in command mode.

jasongrout commented:
Now it seems clearer to me that the modes are notebook modes (like you call it in the model) and the class should be on the notebook widget. For example, we can't have some cells in edit mode and others in command mode. Instead, either the notebook as a whole is in edit mode (editing the current cell) or is in command mode.

blink1073 commented:
We could do that, and have selectors such as `.jp-Notebook.jp-mod-editMode > .jp-CodeCell`.

blink1073 commented:
Then you'd either focus the notebook itself or an editor.

blink1073 commented:
But you have to explicitly focus _something_ other than the codemirror node in order for it to stop editing.

jasongrout commented on src/notebook/model.ts:
```diff
@@ -413,7 +436,7 @@ class NotebookModel implements INotebookModel {
       let cell = this.createCodeCell();
       this.cells.add(cell);
-      cell.mode = 'edit';  // This already sets the new index.
+      this.mode = 'edit';  // This already sets the new index.
```
This no longer sets the new ~~active~~current cell index, right?

blink1073 replied:
I am going to explore this part of the code further in the PR that moves `active` and `selected` off of the cells.

jasongrout commented on src/notebook/widget.ts:
```diff
@@ -384,6 +395,48 @@ class NotebookWidget extends Widget {
+   * Initialize a cell widget.
+   */
+  protected initCell(cell: BaseCellWidget): void {
+    cell.input.editor.node.addEventListener('focus', this, true);
```
Since focus events bubble, we should just add a single event listener to the notebook widget, instead of one to each cell widget.

blink1073 replied:
Presumably we can differentiate by class name on the event target.

jasongrout replied:
Class names aren't part of the interface for cells, are they? Perhaps they should be?

jasongrout commented on src/notebook/widget.ts:
```diff
@@ -391,36 +444,79 @@ class NotebookWidget extends Widget {
-      layout.insertChild(args.newIndex, factory(args.newValue as ICellModel));
+      widget = factory(args.newValue as ICellModel);
+      layout.insertChild(args.newIndex, widget);
+      this.initCell(widget);
```
This case can be reverted to the old code when we just add a single focus/blur handler to the notebook widget.

jasongrout commented:
The logic seems much easier to follow now. I think this is right direction. Thanks for tackling this!

Commit:
More mode handling cleanup

blink1073 commented:
Moved to a single focus/blur handler and focus the notebook itself instead of cells.  

I'd say the class names are public already in that they are defined at the top level of each example file and in the example CSS, I'm not sure how to make them any more public.

jasongrout commented:
@blink1073 - I guess you're right about the class names. At some point, we should write down in docs somewhere what we consider the public interface of a component (e.g., the interface, the css class names, etc.)

Commit:
Tag the cell to simplify the css selector

Commit:
Clean up mode class logic

Commit:
Bail if the focused/blurred cell is found

Commit:
More cleanup of focus/blur handling

blink1073 commented:
@jasongrout, ready for review.

jasongrout commented:
looking at it now...

jasongrout commented on src/notebook/widget.ts:
```diff
@@ -154,6 +154,22 @@ const TOOLBAR_PRESSED = 'jp-mod-pressed';
+ * The class name added to Editor widget instances.
+ * Keep in sync with `editor/widget.ts`.
+ */
+const EDITOR_CLASS = 'jp-Editor';
```
Perhaps we should export this from the editor file? On the other hand, we wouldn't be able to import the const into a CSS file, so we already have to keep things in sync with the CSS files by hand.

blink1073 replied:
In general this is an open problem.

jasongrout commented on src/notebook/widget.ts:
```diff
@@ -391,36 +409,81 @@ class NotebookWidget extends Widget {
+   * Handle `update-request` messages sent to the widget.
+   */
+  protected onUpdateRequest(msg: Message): void {
+    // Set the appropriate classes on the cells.
```
Why are we setting the class redundantly on the cells? I think we should just set the class on the notebook widget, and the css for the cells should depend on having an ancestor in edit or command mode.

jasongrout commented:
This looks great, except for one comment about setting the CSS class on the cells (see the inline comment).

Commit:
Use proper class tagging

blink1073 commented:
@jasongrout, I cleaned up the CSS tagging.  The notebook now gets tags the editors so we don't need to keep the class in sync.  I updated the CSS to use the `jp-Notebook-cell` class.  The reason to use this is that it simplifies the CSS selector, and properly uses semantic tagging.

jasongrout commented on src/notebook/widget.ts:
```diff
@@ -164,10 +164,9 @@ const EDIT_CLASS = 'jp-mod-editMode';
+ * The class name added to notebook editor instances.
  */
-const EDITOR_CLASS = 'jp-Editor';
+const NB_EDITOR_CLASS = 'jp-Notebook-editor';
```
Great idea!

jasongrout commented on src/notebook/widget.ts:
```diff
@@ -201,6 +200,7 @@ class NotebookWidget extends Widget {
       widget = new BaseCellWidget(cell);
     }
     widget.addClass(NB_CELL_CLASS);
+    widget.input.editor.addClass(NB_EDITOR_CLASS);
```
I guess this means that the cell widget can't swap out the editor widget under our noses. Something to keep in mind...

jasongrout replied:
Or really, just that if the editor widget is swapped out, you should copy the classes over.

blink1073 replied:
I moved the addClass logic out of the factory so we preserve the tagging in subclasses.

jasongrout commented on src/notebook/widget.ts:
```diff
@@ -391,36 +409,81 @@ class NotebookWidget extends Widget {
+    if (model.mode === 'edit') {
+      this.addClass(EDIT_CLASS);
+      this.removeClass(COMMAND_CLASS);
+      if (widget) {
```
This was the focus of my comment about setting redundant classes on the cell widgets. I don't think we should set edit and command classes on the cells, only on the notebook widgets. Then the cell CSS can be `.EDIT_CLASS .cell_class` or `.COMMAND_CLASS .cell_class`.

jasongrout commented on src/notebook/widget.ts:
```diff
@@ -391,36 +409,81 @@ class NotebookWidget extends Widget {
+      this.addClass(EDIT_CLASS);
+      this.removeClass(COMMAND_CLASS);
+      if (widget) {
+        widget.addClass(EDIT_CLASS);
```
This was the focus of my comment about setting redundant classes on the cell widgets. I don't think we should set edit and command classes on the cells, only on the notebook widgets. Then the cell CSS can be `.EDIT_CLASS .cell_class` or `.COMMAND_CLASS .cell_class`.

blink1073 replied:
I was under the impression that it would be less efficient to match that way, but it turns out to be about the same performance.

Commit:
Clean up css tagging

jasongrout commented on src/notebook/widget.ts:
```diff
@@ -199,8 +199,6 @@ class NotebookWidget extends Widget {
       widget = new BaseCellWidget(cell);
     }
-    widget.addClass(NB_CELL_CLASS);
-    widget.input.editor.addClass(NB_EDITOR_CLASS);
```
I assume you removed these lines so that it would be easier to override this function, but still have all of the class logic work?

blink1073 replied:
Yep

jasongrout commented:
Awesome. This is great!
